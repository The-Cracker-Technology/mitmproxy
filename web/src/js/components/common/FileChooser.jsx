import React  from 'react'
import PropTypes from 'prop-types'

FileChooser.propTypes = {
    icon: PropTypes.string,
    text: PropTypes.string,
    className: PropTypes.string,
    title: PropTypes.string,
    onOpenFile: PropTypes.func.isRequired
}

export default function FileChooser({ icon, text, className, title, onOpenFile }) {
    let fileInput;
    return (
        <a href='#' onClick={() => fileInput.click()}
           className={className}
           title={title}>
            <i className={'fa fa-fw ' + icon}></i>
            {text}
             <input
                ref={ref => fileInput = ref}
                className="hidden"
                type="file"
                onChange={e => { e.preventDefault(); if(e.target.files.length > 0) onOpenFile(e.target.files[0]); fileInput.value="";}}
            />
        </a>
    )
}
