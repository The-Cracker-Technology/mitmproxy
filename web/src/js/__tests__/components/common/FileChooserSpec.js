import React from 'react'
import renderer from 'react-test-renderer'
import FileChooser from '../../../components/common/FileChooser'

describe('FileChooser Component', () => {
    let openFileFunc = jest.fn(),
        createNodeMock = () => { return { click: jest.fn() } },
        fileChooser = renderer.create(
            <FileChooser className="foo" title="bar" onOpenFile={ openFileFunc }/>
        , { createNodeMock })
        //[test refs with react-test-renderer](https://github.com/facebook/react/issues/7371)

    it('should render correctly', () => {
        let tree = fileChooser.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should handle click action', () => {
        let tree = fileChooser.toJSON(),
            mockEvent = {
                preventDefault: jest.fn(),
                target: {
                    files: [ "foo", "bar" ]
                }
            }
        tree.children[1].props.onChange(mockEvent)
        expect(openFileFunc).toBeCalledWith("foo")
        tree.props.onClick()
        // without files
        mockEvent = {
            ...mockEvent,
            target: { files: [ ]}
        }
        openFileFunc.mockClear()
        tree.children[1].props.onChange(mockEvent)
        expect(openFileFunc).not.toBeCalled()
    })
})
