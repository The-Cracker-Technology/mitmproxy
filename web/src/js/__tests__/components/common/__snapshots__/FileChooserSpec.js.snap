// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileChooser Component should render correctly 1`] = `
<a
  className="foo"
  href="#"
  onClick={[Function]}
  title="bar"
>
  <i
    className="fa fa-fw undefined"
  />
  <input
    className="hidden"
    onChange={[Function]}
    type="file"
  />
</a>
`;
